torch/custom_class.h

```diff
@@ -127,6 +109,63 @@ class class_ {
     return *this;
   }
 
+  // Pickle
+  template <typename GetStateFn, typename SetStateFn>
+  class_& def(detail::pickle_factory<GetStateFn, SetStateFn> pickle) {
+    def("__getstate__", std::move(pickle.g));
+
+    // __setstate__ needs to be registered with some custom handling:
+    // We need to wrap the invocation of of the user-provided function
+    // such that we take the return value (i.e. c10::intrusive_ptr<CurrClass>)
+    // and assign it to the `capsule` attribute.
+    auto s = pickle.s;
+    auto setstate_wrapper = [s](c10::tagged_capsule<CurClass> self,
+                                typename decltype(pickle)::arg_state_type arg) {
+      c10::intrusive_ptr<CurClass> classObj = at::guts::invoke(s, arg);
+      auto genericPtr =
+          c10::static_intrusive_pointer_cast<torch::jit::CustomClassHolder>(
+              classObj);
+      auto capsule = IValue(genericPtr);
+      auto object = self.ivalue.toObject();
+      object->setSlot(0, capsule);
+    };
+    defineMethod<void>("__setstate__", std::move(setstate_wrapper));
+
+    // type validation
+    auto getstate_schema = classTypePtr->getMethod("__getstate__")->getSchema();
+    auto format_getstate_schema = [&getstate_schema]() {
+      std::stringstream ss;
+      ss << getstate_schema;
+      return ss.str();
+    };
+    TORCH_CHECK(
+        getstate_schema.arguments().size() == 1,
+        "__getstate__ should take exactly one argument: self. Got: ",
+        format_getstate_schema());
+    auto first_arg_type = getstate_schema.arguments().at(0).type();
+    TORCH_CHECK(
+        *first_arg_type == *classTypePtr,
+        "self argument of __getstate__ must be the custom class type. Got ",
+        first_arg_type->str());
+    TORCH_CHECK(
+        getstate_schema.returns().size() == 1,
+        "__getstate__ should return exactly one value for serialization. Got: ",
+        format_getstate_schema());
+    auto ser_type = getstate_schema.returns().at(0).type();
+    auto setstate_schema = classTypePtr->getMethod("__setstate__")->getSchema();
+    TORCH_INTERNAL_ASSERT(setstate_schema.arguments().size() == 2);
```

Comment:
shouldn't it be TORCH_CHECK too?

Reply:
This is the number of arguments for the wrapper we generated above, not the user's code, so it wouldn't be meaningful to TORCH_CHECK on that